### PR TITLE
Add TypeScript support

### PR DESCRIPTION
**Pull request by Jeasmine**
#### :page_facing_up: Description:

This PR includes TS support for OneSignal Cordova SDK js plugin interface. For TS Support and last frameworks supports as Capacitor, we need to update our way of exporting the SDK. 

As an improvement module.export has been updated to

`var OneSignal = new OneSignalPlugin();
  module.exports = OneSignal;`

For backward compatibility and windows.plugin import

`if(!window.plugins)   
     window.plugins = {};
  if (!window.plugins.OneSignal)    
     window.plugins.OneSignal = OneSignal;`

 has been kept

#### :pushpin: Notes:

Tested with cordova app and capacitor app, with `window.plugin.Onesignal`, `OneSignal` direct access, and` import OneSignal from 'onesignal-cordova-plugin';`

#### :heavy_check_mark: Performed Tasks:

* Create typing files
* Rename OneSignal.js to OneSignalPlugin.js
* Export typed OneSignal
* Update methods to new names, rename from handleNotificationWillShowInForeground to setNotificationWillShowInForegroundHandler, rename from handleNotificationOpened to setNotificationOpenedHandler, rename from handleInAppMessageClicked to setInAppMessageClickHandler
* Update setLogLevel params
* Add OSSMSSubscriptionObserver to OneSignalPush.h
* Add Notification getNotification method to OSNotificationReceivedEvent
* Update export method to be compatible with TS and window.plugin import

 

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704)

### CODE REVIEW

File: typing/Notification.d.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+export interface OpenedEvent {
+    action          : OpenedEventAction;
+    notification    : OSNotification;
+}
+
+export interface OpenedEventAction {
+    type : OpenedEventActionType
+}
```
Comment (rgomezp) [inline typing/Notification.d.ts:53]: newline

File: typing/OneSignalPlugin.d.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+     */
+    setRequiresUserPrivacyConsent(required: boolean): void;
+
+    /**
+     * If your application is set to require the user's privacy consent, you can provide this consent using this method.
+     * @param  {boolean} granted
+     * @returns void
+     */
+    provideUserConsent(granted: boolean): void;
+}
```
Comment (rgomezp) [inline typing/OneSignalPlugin.d.ts:338]: newline

File: www/OneSignalPlugin.js [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+
+OneSignalPlugin.prototype.isLocationShared = function(callback) {
+    window.cordova.exec(callback, function() {}, "OneSignalPush", "isLocationShared", []);
+};
+
+//-------------------------------------------------------------------
+
+var OneSignal = new OneSignalPlugin();
+
+module.exports = OneSignal;
```
Comment (rgomezp) [inline www/OneSignalPlugin.js:454]: newline

File: typing/OneSignalPlugin.d.ts [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    /**
+     * Removes whatever was set as the current user's external user ID.
+     * @param  {(results:object)=>void} handler
+     * @returns void
+     */
+    removeExternalUserId(handler?: (results: object) => void): void;
+
+    /**
+     * Sets an In-App Message click event handler.
+     * @param  {(action:InAppMessageAction)=>void} handler
```
Comment (rgomezp) [inline typing/OneSignalPlugin.d.ts:225]: missing import for `InAppMessageAction`

File: typing/Notification.d.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    attachments         ?: object;
+    mutableContent      ?: boolean;
+    contentAvailable    ?: string;
+}
+
+/* N O T I F I C A T I O N   &   I A M   E V E N T S */
+export interface NotificationReceivedEvent {
+    complete        : (notification?: OSNotification) => void;
+    getNotification : () => OSNotification;
+};
```
Comment (rgomezp) [inline typing/Notification.d.ts:44]: remove semicolon

File: typing/Subscription.d.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    notificationPermissionStatus    ?: IosPermissionStatus;  // ios only
+    // areNotificationsEnabled (android) not included since it is converted to hasNotificationPermission in bridge
+}
+
+export interface PermissionChange {
+    status                  ?: IosPermissionStatus;    // ios
+    hasPrompted             ?: boolean;   // ios
+    provisional             ?: boolean;   // ios
+    areNotificationsEnabled ?: boolean;   // android
+};
```
Comment (rgomezp) [inline typing/Subscription.d.ts:26]: remove semi-colon

File: typing/Subscription.d.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+export interface SubscriptionChange {
+    userId                  ?: string;
+    pushToken               ?: string;
+    isSubscribed            : boolean;
+    isPushDisabled          : boolean;
+};
```
Comment (rgomezp) [inline typing/Subscription.d.ts:33]: remove semi-colon

File: typing/Subscription.d.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+export interface EmailSubscriptionChange {
+    emailAddress        ?: string;
+    emailUserId         ?: string;
+    isEmailSubscribed   : boolean;
+};
```
Comment (rgomezp) [inline typing/Subscription.d.ts:39]: remove semi-colon

File: typing/Subscription.d.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+export interface SMSSubscriptionChange {
+    smsNumber         ?: string;
+    smsUserId         ?: string;
+    isSMSSubscribed   : boolean;
+};
```
Comment (rgomezp) [inline typing/Subscription.d.ts:45]: remove semi-colon

File: typing/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { OSNotification, NotificationReceivedEvent, OpenedEvent, OpenedEventAction, OpenedEventActionType } from './Notification';
+import { OutcomeEvent } from './Outcomes';
+import { PermissionChange, SubscriptionChange, EmailSubscriptionChange, SMSSubscriptionChange, DeviceState, IosPermissionStatus } from './Subscription';
+import { LogLevel, ChangeEvent } from './Extras';
+import { OneSignalPlugin } from './OneSignalPlugin';
+export declare const OneSignal: OneSignalPlugin;
+export { OneSignalPlugin };
+export default OneSignal;
```
Comment (rgomezp) [inline typing/index.d.ts:8]: newline

File: typing/OneSignalPlugin.d.ts (hunk 4 above)
Comment (rgomezp) [reply typing/OneSignalPlugin.d.ts:225]: also missing definition for it

**Comment by rgomezp** [comment]
Let's rename `typing` to `types`

File: typing/OneSignalPlugin.d.ts [hunk 11]
```diff
@@ -0,0 +1,338 @@
+import { NotificationReceivedEvent, OpenedEvent, OpenedEventAction } from './Notification';
+import { OutcomeEvent } from './Outcomes';
+import { PermissionChange, SubscriptionChange, EmailSubscriptionChange, SMSSubscriptionChange, DeviceState } from './Subscription';
+import { LogLevel, ChangeEvent } from './Extras';
+
+/* O N E S I G N A L  I N T E R F A C E */
+export interface OneSignalPlugin{
```
Comment (rgomezp) [inline typing/OneSignalPlugin.d.ts:7]: missing space before {

File: typing/index.d.ts (hunk 10 above)
Comment (rgomezp) [reply typing/index.d.ts:8]: I'm not entirely sure we should be exporting `OneSignal` as default here. The plugin is accessed via the plugins array (`window.plugins.OneSIgnal`). So to add support, I think we need to instruct users to use something like:

```
const OneSignal: OneSignalPlugin = window["plugins"].OneSignal;
// ...
```

In other words, we allow them to import the types via the `OneSignalPlugin` interface which they then have to declare as the type for the plugin object itself.

I can't really think of a better way and I haven't been able to find too many resources on how best to handle this scenario.

**Review by rgomezp — CHANGES_REQUESTED** [review]

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704)* status: 4 of 9 files reviewed, 17 unresolved discussions (waiting on @Jeasmine and @rgomezp)

___
*[package.json, line 8 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKWKnK7gOVSA6LD8MU:-MbKWKnK7gOVSA6LD8MV:b4rh5fv) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/package.json#L8)):*
> ```JSON
>   "description": "OneSignal is a high volume Push Notification service for mobile apps. In addition to basic notification delivery, OneSignal also provides tools to localize, target, schedule, and automate notifications that you send.",
>   "license": "MIT",
>   "main": "www/OneSignal.js",
> ```

This file was renamed to `OneSignalPlugin.js` so we have to update the name here.
___
*[package.json, line 31 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKLmpU613m9wRBaoi-:-MbKLmpU613m9wRBaoi0:b-snu8ii) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/package.json#L31)):*
> ```JSON
>   ],
>   "dependencies": {
>     "typescript": "^4.1.3"
> ```

Wouldn't this add the Typescript dependency to those who don't even have it though? We don't have this in the [react-native-onesignal/package.json](https://github.com/OneSignal/react-native-onesignal/blob/main/package.json)
___
*[typing/index.d.ts, line 1 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKVaXW3nG00-ao_PLZ:-MbKVaXW3nG00-ao_PL_:bnsyn51) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/index.d.ts#L1)):*
> ```XML
> import { OSNotification, NotificationReceivedEvent, OpenedEvent, OpenedEventAction, OpenedEventActionType } from './Notification';
> ```

Do we need to define each file here? If we just import `OneSignalPlugin` does it not import everything in it automatically? If we do need this then can we import `*` from these files so we don't have to maintain this file everytime we add a new class to any of it's child files?
___
*[typing/index.d.ts, line 8 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohY-r1-8:-MbKU6bz0RoXdFwy0_Ar:b-m8fsgk) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/index.d.ts#L8)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

I'm not entirely sure we should be exporting `OneSignal` as default here. The plugin is accessed via the plugins array (`window.plugins.OneSIgnal`). So to add support, I think we need to instruct users to use something like:

```
const OneSignal: OneSignalPlugin = window["plugins"].OneSignal;
// ...
```

In other words, we allow them to import the types via the `OneSignalPlugin` interface which they then have to declare as the type for the plugin object itself.

I can't really think of a better way and I haven't been able to find too many resources on how best to handle this scenario.
</blockquote></details>

I think since the following lines are now at the bottom of `www/OneSignalPlugin.js` this might be correct:

```js
var OneSignal = new OneSignalPlugin();
module.exports = OneSignal;
```
___
*[typing/Subscription.d.ts, line 16 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKQ438D1WSjNoHM7i4:-MbKQ438D1WSjNoHM7i5:b-sd8axz) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L16)):*
> ```XML
>     isEmailSubscribed               : boolean;
>     isSMSSubscribed                 : boolean;
>     hasNotificationPermission       ?: boolean; // ios only
> ```

remove "ios only" comment as `hasNotificationPermission` works for Android too.
___
*[www/OneSignal.js, line 457 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKNEH6D1BE_yB-jjlO:-MbKNEH6D1BE_yB-jjlP:bfcmpoz) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L457)):*
> ```JavaScript
> 
> if (!window.plugins.OneSignal)
>     window.plugins.OneSignal = new OneSignal();
> ```

Removing this is more than a Typescript change, or was this required to get Typescript working?

App developers won't be able to call `window.plugins.OneSignal` any more, however since this is a major release this seems like a good time to push developer to use the more modern`import` syntax. Would be good to know however if there are any specific requirements are needed from the app project are needed or if any more recent Cordova or Ionic project will work just fine.

From a quick look at another Cordova plugin ([fechanique/cordova-plugin-fcm](https://github.com/fechanique/cordova-plugin-fcm)) this seems like it probably is the correct approach but it would be good to document the reasoning and how this works in the PR.
___
*[www/OneSignal.js, line 459 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKPdZe8QnkCJUJgQQo:-MbKPdZe8QnkCJUJgQQp:b-lz1s0d) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L459)):*
> ```JavaScript
>     window.plugins.OneSignal = new OneSignal();
> 
> if (typeof module != 'undefined' && module.exports) {
> ```

Could module be `undefined` for any reason? This line is very old so it might not but we should confirm.

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704)* status: 0 of 15 files reviewed, 17 unresolved discussions (waiting on @jkasten2 and @rgomezp)

___
*[package.json, line 8 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKWKnK7gOVSA6LD8MU:-McB18B8A86SDtWNqkoz:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/package.json#L8)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

This file was renamed to `OneSignalPlugin.js` so we have to update the name here.
</blockquote></details>

Done.
___
*[package.json, line 31 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKLmpU613m9wRBaoi-:-McB1Apo5_kL3sdxL06k:b9dynvy) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/package.json#L31)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Wouldn't this add the Typescript dependency to those who don't even have it though? We don't have this in the [react-native-onesignal/package.json](https://github.com/OneSignal/react-native-onesignal/blob/main/package.json)
</blockquote></details>

Not needed, good catch!
___
*[typing/index.d.ts, line 1 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKVaXW3nG00-ao_PLZ:-McB2Szf6MrZ1XwXnyHX:bryl2o5) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/index.d.ts#L1)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Do we need to define each file here? If we just import `OneSignalPlugin` does it not import everything in it automatically? If we do need this then can we import `*` from these files so we don't have to maintain this file everytime we add a new class to any of it's child files?
</blockquote></details>

No need, good catch!
___
*[typing/index.d.ts, line 8 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohY-r1-8:-McB2EtZ1tJHbIlHq3oh:b-xjoim8) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/index.d.ts#L8)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

I think since the following lines are now at the bottom of `www/OneSignalPlugin.js` this might be correct:

```js
var OneSignal = new OneSignalPlugin();
module.exports = OneSignal;
```
</blockquote></details>

I have been able to test different types of imports, from my point of view this seems to be the more stand and updated one. 
I tested this with a Capacitor app, I was able to import OneSignal and access the TS file with autocomplete, but not able to do so from an Cordova app
___
*[typing/Notification.d.ts, line 44 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohU-r1-44:-McB1DcwElDhKFHF1Q1m:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Notification.d.ts#L44)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

remove semicolon
</blockquote></details>

Done.
___
*[typing/Notification.d.ts, line 53 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohU-r1-53:-McB1E3F2XtAhYIhBSaN:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Notification.d.ts#L53)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.
___
*[typing/OneSignalPlugin.d.ts, line 7 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohV-r1-7:-McB1Ee3A4aMUemlBydA:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/OneSignalPlugin.d.ts#L7)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

missing space before {
</blockquote></details>

Done.
___
*[typing/OneSignalPlugin.d.ts, line 338 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohV-r1-338:-McB1FBQ4tvZjCBFHOPZ:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/OneSignalPlugin.d.ts#L338)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.
___
*[typing/Subscription.d.ts, line 16 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKQ438D1WSjNoHM7i4:-McB1GM167wA6YmbVEbj:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L16)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

remove "ios only" comment as `hasNotificationPermission` works for Android too.
</blockquote></details>

Done.
___
*[typing/Subscription.d.ts, line 26 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohX-r1-26:-McB1Gw7CwgsnlB7TYin:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L26)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

remove semi-colon
</blockquote></details>

Done.
___
*[typing/Subscription.d.ts, line 33 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohX-r1-33:-McB1HGc4-Mxgfqx_BaF:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L33)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

remove semi-colon
</blockquote></details>

Done.
___
*[typing/Subscription.d.ts, line 39 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohX-r1-39:-McB1HcrBKgvQFebtNrh:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L39)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

remove semi-colon
</blockquote></details>

Done.
___
*[typing/Subscription.d.ts, line 45 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohX-r1-45:-McB1I-L--Cku1ETdtc5:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/Subscription.d.ts#L45)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

remove semi-colon
</blockquote></details>

Done.
___
*[www/OneSignal.js, line 457 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKNEH6D1BE_yB-jjlO:-McB1NMC2av3Wk9gViDw:bww9ukt) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L457)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Removing this is more than a Typescript change, or was this required to get Typescript working?

App developers won't be able to call `window.plugins.OneSignal` any more, however since this is a major release this seems like a good time to push developer to use the more modern`import` syntax. Would be good to know however if there are any specific requirements are needed from the app project are needed or if any more recent Cordova or Ionic project will work just fine.

From a quick look at another Cordova plugin ([fechanique/cordova-plugin-fcm](https://github.com/fechanique/cordova-plugin-fcm)) this seems like it probably is the correct approach but it would be good to document the reasoning and how this works in the PR.
</blockquote></details>

After investigating and testing different approaches, I came up with a configuration that supports different types of imports, please check the updates description
___
*[www/OneSignal.js, line 459 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKPdZe8QnkCJUJgQQo:-McB1Y7n1K35QPqhmfgC:b-mbhgg5) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L459)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Could module be `undefined` for any reason? This line is very old so it might not but we should confirm.
</blockquote></details>

If I understand this right, this should not be null or undefined, also couldn't find a source that said it could be undefined. Maybe was an old check?
___
*[www/OneSignalPlugin.js, line 454 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohZ-r1-454:-McB1IVa-ImN--LvmwFJ:b-896fix) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/www/OneSignalPlugin.js#L454)):*
<details><summary><i>Previously, rgomezp (Rodrigo Gomez Palacio) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.

**Review by jkasten2 — APPROVED** [review]
> Reviewed 16 of 16 files at r2.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704)* status: all files reviewed, 12 unresolved discussions (waiting on @Jeasmine and @rgomezp)

___
*[typing/index.d.ts, line 8 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohY-r1-8:-McI-nXl8GMaccXbOj3o:b-8f2cgj) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/typing/index.d.ts#L8)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

I have been able to test different types of imports, from my point of view this seems to be the more stand and updated one. 
I tested this with a Capacitor app, I was able to import OneSignal and access the TS file with autocomplete, but not able to do so from an Cordova app
</blockquote></details>

I think standard Cordova doesn't have TypeScript by default so this makes sense. So this looks good to me.
___
*[types/Notification.d.ts, line 10 at r2](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-McHOXSQ4rFHjONJfAiG:-McHOXSQ4rFHjONJfAiH:b7a2tlm) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/ea1b43ca3e77ea838bd9e5d58ec3e180e388beb4/types/Notification.d.ts#L10)):*
> ```XML
>     title           ?: string;
>     launchURL       ?: string;
>     rawPayload      : object | string; // platform bridges return different types
> ```

Would be good to fix in this in a follow up beta3.
___
*[www/OneSignal.js, line 457 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKNEH6D1BE_yB-jjlO:-McHPhymDBvNan8a3Knv:bh5wghn) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L457)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

After investigating and testing different approaches, I came up with a configuration that supports different types of imports, please check the updates description
</blockquote></details>

Ah, good idea we can support both ways for an easier upgrade for developers.
___
*[www/OneSignal.js, line 459 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-MbKPdZe8QnkCJUJgQQo:-McHy44K8SwK41Zd06zF:b3z3hdi) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/86b1dd0ff913eeb01f3e5f739b237cd4a0c97b56/www/OneSignal.js#L459)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

If I understand this right, this should not be null or undefined, also couldn't find a source that said it could be undefined. Maybe was an old check?
</blockquote></details>

I wasn't able to find any official documentation about`window.plugins`.

However I did find some documentation on `cordova.require` which seems like the newer way to do this.
https://cordova.apache.org/docs/en/latest/plugin_ref/spec.html

But I also see we have always used `<clobbers ...>` in our `plugin.xml` so it seems to me that Cordova will make `window.OneSignal` accessible.
https://cordova.apache.org/docs/en/latest/plugin_ref/spec.html#clobbers

Anyway let's leave `window.plugins` as-is in this PR. Something we can look at for Beta3
___
*[www/OneSignalPlugin.js, line 454 at r1](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704#-Mb7HQKw1kSzs7p4gohZ-r1-454:-McI-Bog3rNV1adkMjJi:by8lps7) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/babcc7137702204926cf6e628c2e5bec7e472b79/www/OneSignalPlugin.js#L454)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit, looks like this got re-added.

**Review by jkasten2 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/704)* status: all files reviewed, 11 unresolved discussions (waiting on @rgomezp)

**Pull request merged**